Add NotificationService for rental

azzilian opened the pull request:
Add NotificationService & NotificationServiceImpl - Telegram sendMessage - with stub commands
Add stubs - Rental enity, Rental repository, Rental service, rental table creation
Add BotConfig for telegram bot
Add dependency for telegram bot

nklimovych committed:
Initial commit

azzilian committed:
AppConfig.java - added Bean for BotController for registration
BotService.java - register user + send message about user
Liquibase - raw creation table + users insert
UserRepository - existByChatId + findByChatId
UserService - search user in DB + search chatId
Raw User entity - need to update
pom.xml - telegram dependency added

azzilian committed:
AppConfig.java - fixes to be able to read enviroment variables
BotService.java - handlers added

azzilian committed:
Refactor - AppConfig.java -> BotConfig + code change
Add - NotificationService.java
Refactor - NotificationServiceImpl.java - added new functional
Refactor - db.changelog-master.yaml - stub
Add - Rental,RentalService - stub
Removed - liquibase users creation stubs

SkillfulDev committed:
Add - Authentication with JWT

azzilian committed:
Refactor - RentalRepository.java - stub and querry for returnDate
Refactor - UserRepository - stub
Refactor - pom.xml - EV now could be gained from .env
Refactor - NotificationServiceImpl - now takes EV from .env file + new logic to send Status info about returnDate
Remove - application.properties - EV removed

azzilian committed:
Refactor - NotificationServiceImpl - reading EV from application properties
Remove - pom.xml - revomed dependency for reading from .env

SkillfulDev committed:
Refactor - change RoleName to RoleType

nklimovych commented on pom.xml:
```diff
@@ -120,6 +120,17 @@
+            <artifactId>dotenv-java</artifactId>
+            <version>2.2.0</version>
+        </dependency>
+
```
```suggestion
```

nklimovych commented on src/main/java/online/carsharing/config/BotConfig.java:
```diff
@@ -0,0 +1,25 @@
+@ComponentScan("online.carsharing")
+@RequiredArgsConstructor
+public class BotConfig {
+
```
```suggestion
```

nklimovych commented on src/main/java/online/carsharing/repository/RentalRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface RentalRepository extends JpaRepository<Rental, Long> {
+    @Query("SELECT r.returnDate FROM Rental r WHERE r.id = :id")
+    LocalDate returnDateCheckById(Long id);
```
```suggestion
    LocalDate findReturnDateById(Long id);
```

nklimovych commented on src/main/java/online/carsharing/service/NotificationService.java:
```diff
@@ -0,0 +1,9 @@
+public interface NotificationService {
+    void returnStatus(Update update, Long id);
+
+    void sendMessage(Update update, String text);
```
```suggestion
    void sendMessage(Update update, String message);
```

nklimovych commented on src/main/java/online/carsharing/service/impl/NotificationServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+        return this.botToken;
+    }
+
+    @Async
+    @Override
+    public void onUpdateReceived(Update update) {
+        if (update.hasMessage() && update.getMessage().hasText()) {
+            String command = update.getMessage().getText().trim();
+
+            if (command.startsWith("/returnStatus")) {
+                String[] parts = command.split(" ");
+                if (parts.length == 2) {
+                    try {
+                        Long id = Long.parseLong(parts[1]);
+                        returnStatus(update, id);
+                    } catch (NumberFormatException e) {
+                        sendMessage(update, "Invalid ID format. Please use a number.");
+                    }
+                } else {
+                    sendMessage(update, "Invalid command format. Please use /returnStatus <ID>.");
+                }
+            } else {
+                sendMessage(update, "Unknown command");
+            }
+        }
+    }
```
I believe it's better to avoid using magic numbers

nklimovych commented on src/main/java/online/carsharing/service/impl/NotificationServiceImpl.java:
```diff
@@ -0,0 +1,90 @@
+        if (update.hasMessage() && update.getMessage().hasText()) {
+            String command = update.getMessage().getText().trim();
+
+            if (command.startsWith("/returnStatus")) {
```
```suggestion
            if (command.startsWith("/status")) {
```

nklimovych reviewed (commented):
I left you some minor suggestions. I believe we should first merge the rentals and then merge this branch.

azzilian committed:
Remove - Rental completely
Add - Car with stubs for controller
Refactor - NotificationService added method with carCreation with stub

azzilian committed:
Add - UserChatId - entity to connect users and chats
BugFix - CarRepository - correct type returns
Remove - CarController - authoriation removed - DO NOT MERGE
Refactor - NotificationServiceImpl.java -now send message to all users with chats with bot

HAIOVYI commented on src/main/java/online/carsharing/entity/UserChatId.java:
```diff
@@ -0,0 +1,21 @@
+@Setter
+public class UserChatId {
+    @Id
+    @Column(name = "user_id", nullable = false)
+    private Long userId;
+
+    @Column(name = "chat_id", nullable = false)
+    private Long chatId;
```
General question to the team, do we somehow name the columns in this way?

azzilian replied:
Good point I can change user_id into user_telegram_id

HAIOVYI commented on src/main/java/online/carsharing/repository/user/UserChatIdRepository.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.data.jpa.repository.Query;
+
+public interface UserChatIdRepository extends JpaRepository<UserChatId, Long> {
+    @Query("SELECT uc.chatId FROM UserChatId uc")
+    List<Long> findAllChatIds();
+    
+    @Query("SELECT uc FROM UserChatId uc WHERE uc.userId = :userId")
+    UserChatId findByUserId(Long userId);
```
Is it possible to make this query using spring build method? Because it seems possible

azzilian committed:
Bugfix - empty line removed + non-used peendency removed

HAIOVYI commented on src/main/java/online/carsharing/service/impl/NotificationServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+            if (command.startsWith("/addMe")) {
+                addUserChatId(update);
+            } else if (command.startsWith("/getModel")) {
+                String[] parts = command.split(" ");
```
```suggestion
                String[] parts = command.split(CONSTANT);

```

azzilian committed:
Removed command /returStatus - will be added in other task

HAIOVYI commented on src/main/java/online/carsharing/service/impl/NotificationServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+                String[] parts = command.split(" ");
+                if (parts.length == 2) {
+                    try {
+                        Long id = Long.parseLong(parts[1]);
```
magic numbers

azzilian replied:
I will change it but right now this functional moved to other branch

HAIOVYI commented on src/main/java/online/carsharing/service/impl/NotificationServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+            }
+        } catch (Exception e) {
+            sendMessage(update, "Error fetching car info");
+            e.printStackTrace();
```
I don't know exactly, is it ok to use printStackTrace() here?

HAIOVYI commented on src/main/java/online/carsharing/service/impl/NotificationServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+                : update.getCallbackQuery().getFrom().getId();
+        try {
+            execute(SendMessage.builder().chatId(chatId.toString()).text(text).build());
+        } catch (TelegramApiException e) {
```
the same here

HAIOVYI commented on src/main/java/online/carsharing/service/impl/NotificationServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+                            .text(messageText)
+                            .build());
+                } catch (TelegramApiException e) {
+                    e.printStackTrace();
```
same

HAIOVYI commented on src/main/resources/db/changelog/changes/vyshynskyi/create-users-chat_id-table.yaml:
```diff
@@ -0,0 +1,18 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-users-roles-table
+      author: vyshynskyi
+      changes:
+        - createTable:
+            tableName: users_chat_id
+            columns:
+              - column:
+                  name: user_id
+                  type: bigint
+                  constraints:
+                    nullable: false
+              - column:
+                  name: chat_id
+                  type: bigint
+                  constraints:
+                    nullable: false
```
Table without primary key?

azzilian replied:
Same as users-roles table, well at least for now there is no need on primary key - because this PR is only for send messages

azzilian committed:
Update - NotificationServiceImpl.java - response style update

HAIOVYI reviewed (changes requested):
Good job, but I think we need to discuss about this PR

azzilian committed:
Add - test bot for est enviroment

azzilian committed:
Added Stub classes for Rental creation

azzilian committed:
Added find by Id for user
Refactor NotificationServiceImpl - proper email responce

azzilian commented:
-    Removed - stubs classes, though RentalControler still have stub method

-     Regarding comments on Exception magic numbers and requests to the repository - I'll upload them as separate fixes, creating a Trello task for this.

-     Table entity UserChatId - for now, let it be until I figure out how to implement authorization via Telegram.
 For the main task of sending messages, it works.

azzilian committed:
Removed stubs
Add RentalRepository
Refactor RentalController
Add CarsNotAvailableException
Refactor File Structure
Refactor Mapper - responsedto

azzilian committed:
Merge branch main into telegram-implementation

azzilian committed:
Refactor RentalRequestDto - Removed @NotBlack

azzilian committed:
Bugfix fixed bug n carcontroller

azzilian committed:
Merge remote-tracking branch 'origin/main' into telegram-implementation

# Conflicts:
#	src/main/java/online/carsharing/dto/request/rental/RentalRequestDto.java
#	src/main/java/online/carsharing/entity/Car.java
#	src/main/java/online/carsharing/service/impl/RentalServiceImpl.java
#	src/main/resources/db/changelog/db.changelog-master.yaml

azzilian committed:
Refactor:
Car - @ToString addedtelegram
CarServiceImpl - notification to  about car creation added
RentalServiceImpl - notification to  about rental creation added
NotificationService - methods renamed
NotificationServiceImpl - add new string registration message

azzilian committed:
Refactor liquibase primaryKey added

HAIOVYI reviewed (approved):
GJ!
